File: Common_Concepts/bubbleSort.js
Array.prototype.bubbleSort = function (cb) {
  if (typeof cb != 'function') cb = compA;

  let sorted = false;
  while (sorted == false) {
    sorted = true;
    for(let i = 0; i < this.length - 2; i++) {
      if (cb(this[i], this[i + 1]) >= 0) {
        [this[i], this[i + 1] = this[i + 1], this[i]];
        sorted = false;
      };
    };
  };
  return this;
};

const compA = (a, b) => {
  if (a > b) {
    return 1;
  } else if (a == b) {
    return 0;
  } else {
    return -1;
  };
};

const compB = (a, b) => {
 if (a > b) {
   return -1;
 } else if (a == b) {
   return 0;
 } else {
   return 1;
 };
};

// console.log(pojo instanceof Object);
console.log([1,6,3,8,6,5,10].bubbleSort()); // [1,3,5,6,6,8,10]
console.log([1,6,3,8,6,5,10].bubbleSort(compA)); // [10,8,6,6,5,3,1]
console.log([1,6,3,8,6,5,10].bubbleSort(compB)); // [1,3,5,6,6,8,10]

// Array.prototype.bubbleSort = () => {
//   console.log(this);
//   // console.log(`yeeee ${this.what}`);
//   let sorted = false;
//   while (sorted !== true) {
//     sorted = true;
//     for (let i = 0; i < this.what.length - 2; i++) {
//       if (this.what[i] > this.what[i + 1]) {
//         [this.what[i], this.what[i + 1]] = [this.what[i + 1], this.what[i]];
//         sorted = false;
//       }
//     }
//     sorted;
//   }
//   return this.what;
// };

// console.log(Array.prototype.bubbleSort.call([1,6,3,8,6,5,10]));

File: Common_Concepts/bubble_sort.rb
require 'byebug'

# class Array
#   def bubble_sort(&prc)
#     # debugger
#     prc ||= proc { |x, y| x <=> y }
#
#     sorted = false
#     while sorted == false
#       sorted = true
#       (0..length).each do |id|
#         break if id + 1 >= length
#         if prc.call(self[id],self[id + 1]) == -1
#           self[id], self[id + 1] = self[id + 1], self[id]
#           sorted = false
#         end
#       end
#       sorted
#     end
#     self
#   end
# end

class Array
  def bubble_sort(&prc)
    # debugger
    prc = proc { |x, y| x <=> y } if prc == nil

    sorted = false
    while sorted == false
      sorted = true
      (0..self.length - 2).each do |id|
        if prc.call(self[id], self[id + 1]) == -1
          self[id], self[id + 1] = self[id + 1], self[id]
          sorted = false
        end
      end
      sorted
    end
    self
  end
end

p [1, 2, 3, 4, 5, 6, 7].bubble_sort
p [1, 2, 3, 4, 5, 6, 7].bubble_sort { |x, y| y <=> x }
p [7, 1, 8, 3, 0, 1].bubble_sort { |x, y| x <=> y }
